design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+# Contour Operator Gateway API Support Migration
+
+Status: Draft
+
+## Abstract
+Currently the Contour Operator and Contour Controller have both started implementing Gateway API support.
+Both components contain controllers that reconcile GatewayClass and Gateway resources and that conflict when reconciling status.
+This document describes options for how we should proceed to untangle these.
+
+## Background
+Currently the state of the world is as below:
+
+### Contour Operator
+- Configured with a GatewayClass object reference (name) to reconcile
+- Includes GatewayClass controller
+  - Checks if named GatewayClass has the correct (hardcoded to `projectcontour.io/contour-operator`) controller string
+  - Checks for Contour CRD in the parameters ref of the GatewayClass, uses parameters for provisioning Contour instances
+  - Sets admitted status on object if all valid
+- Includes Gateway controller
+  - Checks if Gateway refers to correct GatewayClass and that GatewayClass has been admitted
+  - Sets up finalizer for Gateway
+  - Creates Contour deployment resources with the Gateway being reconciled named in Contour's ConfigMap
+  - Sets up finalizer for GatewayClass
+  - Sets up finalizer for Contour CRD referenced in GatewayClass
+  - Sets admitted status on object if all valid
+
+### Contour Controller
+- Needs to be configured with a GatewayClass controller name to ensure it reconciles the correct GatewayClass
+- Also needs a Gateway namespace/name which is being deprecated
+- Includes GatewayClass controller
+  - Expects parameter ref field to be unset, otherwise GatewayClass is not admitted
+  - Admits GatewayClass that has the configured controller name
+- Includes Gateway controller
+  - Checks if Gateway references Contour owned GatewayClass
+  - Passes off reconciliation to dag package Gateway API processor
+  - Gateway is validated, status set on it and relevant *Routes
+
+### Current Problems
+- Contour requires Gateway namespace/name *and* controller name, the latter of which the operator does not provide
+- Contour expects GatewayClass parameters ref to be unset while Operator expects it to reference a Contour CRD
+- If the above is rectified, Contour and the Operator still will conflict on setting status
+- CI failures that occur because of conflicts between Contour and the Operator are not detected unless we manually look at the scheduled Operator CI job
+
+## Goals
+- Improve implementation in Contour and Operator to mitigate current issues
+- Improve development processes and practices to ensure this conflicting state does not occur again
+
+## Non Goals
+- Drop support for features or projects
+
+## Remediation Steps
+
+- Set up alerting on Operator CI failures
```

Comment:
Can we get CI up to test installation and run e2e for the Contour Operator nightly?

Reply:
Already has been running hourly. Last we we added Slack notifications for updates to when it fails.

Reply:
got it, but we didn't have it prior to the broken release right?

Reply:
We did have it, but I didn't check it. Surfacing the results in the Slack channel should help to avoid others missing it like I did; there's probably also some prerelease items to add to the release process docs to remind us as well.